DataFormats/HcalRecHit/interface/HBHEChannelInfo.h

```diff
@@ -0,0 +1,213 @@
+        {return (rawCharge_[ts] - pedestal_[ts])*gain_[ts];}
+    inline double tsGain(const unsigned ts) const {return gain_[ts];}
+    inline uint8_t tsAdc(const unsigned ts) const {return adc_[ts];}
+    inline float tsRiseTime(const unsigned ts) const
+        {return hasTimeInfo_ ? riseTime_[ts] : HcalSpecialTimes::UNKNOWN_T_NOTDC;}
+
+    // Signal rise time measurement for the SOI, if available
+    inline float soiRiseTime() const
+    {
+        return (hasTimeInfo_ && soi_ < nSamples_) ?
+                riseTime_[soi_] : HcalSpecialTimes::UNKNOWN_T_NOTDC;
+    }
+
+    // The TS with the "end" index is not included in the window
+    inline double chargeInWindow(const unsigned begin, const unsigned end) const
+    {
+        double sum = 0.0;
+        const unsigned imax = end < nSamples_ ? end : nSamples_;
+        for (unsigned i=begin; i<imax; ++i)
+            sum += (rawCharge_[i] - pedestal_[i]);
+        return sum;
+    }
+
+    inline double energyInWindow(const unsigned begin, const unsigned end) const
+    {
+        double sum = 0.0;
+        const unsigned imax = end < nSamples_ ? end : nSamples_;
+        for (unsigned i=begin; i<imax; ++i)
+            sum += (rawCharge_[i] - pedestal_[i])*gain_[i];
+        return sum;
+    }
+
+    // The two following methods return MAXSAMPLES if the specified
+    // window does not overlap with the samples stored
+    inline unsigned peakChargeTS(const unsigned begin, const unsigned end) const
+    {
+        unsigned iPeak = MAXSAMPLES;
+        double dmax = -DBL_MAX;
+        const unsigned imax = end < nSamples_ ? end : nSamples_;
+        for (unsigned i=begin; i<imax; ++i)
+        {
+            const double q = rawCharge_[i] - pedestal_[i];
+            if (q > dmax)
+            {
+                dmax = q;
+                iPeak = i;
+            }
+        }
+        return iPeak;
+    }
+
+    inline unsigned peakEnergyTS(const unsigned begin, const unsigned end) const
+    {
+        unsigned iPeak = MAXSAMPLES;
+        double dmax = -DBL_MAX;
+        const unsigned imax = end < nSamples_ ? end : nSamples_;
+        for (unsigned i=begin; i<imax; ++i)
+        {
+            const double e = (rawCharge_[i] - pedestal_[i])*gain_[i];
+            if (e > dmax)
+            {
+                dmax = e;
+                iPeak = i;
+            }
+        }
+        return iPeak;
+    }
+
+    // The following function can be used, for example,
+    // in a check for presence of saturated ADC values
+    inline uint8_t peakAdcValue(const unsigned begin, const unsigned end) const
+    {
+        uint8_t peak = 0;
+        const unsigned imax = end < nSamples_ ? end : nSamples_;
+        for (unsigned i=begin; i<imax; ++i)
+            if (adc_[i] > peak)
+                peak = adc_[i];
+        return peak;
+    }
+
+private:
+    HcalDetId id_;
+
+    // Charge in fC for all time slices
+    double rawCharge_[MAXSAMPLES];
```

Comment:
Not to check for array bounds is a conscious design decision. The comments
in the header explain that. This class will be used in the HLT, so speed takes
priority.

________________________________
From: David Lange [notifications@github.com]
Sent: Thursday, July 14, 2016 3:34 AM
To: cms-sw/cmssw
Cc: Volobouev, I; Mention
Subject: Re: [cms-sw/cmssw] Initial implementation of the HB/HE Phase 1 reconstruction (#15092)


In DataFormats/HcalRecHit/interface/HBHEChannelInfo.h<https://github.com/cms-sw/cmssw/pull/15092#discussion_r70768187>:

> +    // in a check for presence of saturated ADC values
> +    inline uint8_t peakAdcValue(const unsigned begin, const unsigned end) const
> +    {
> +        uint8_t peak = 0;
> +        const unsigned imax = end < nSamples_ ? end : nSamples_;
> +        for (unsigned i=begin; i<imax; ++i)
> +            if (adc_[i] > peak)
> +                peak = adc_[i];
> +        return peak;
> +    }
> +
> +private:
> +    HcalDetId id_;
> +
> +    // Charge in fC for all time slices
> +    double rawCharge_[MAXSAMPLES];


@igv4321<https://github.com/igv4321> - especially as these variables are exposed via the class interface, they should be std::array to allow for easy bounds checking. Thanks

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/cms-sw/cmssw/pull/15092/files/cb95544ae57dc7aa6cab5481015b80942b61d6d8#r70768187>, or mute the thread<https://github.com/notifications/unsubscribe/AFS4-R6K657fOII9-3Tstuc1Fqa663neks5qVfR7gaJpZM4JCc97>.